modern-publish-plugin.md

```diff
@@ -0,0 +1,222 @@
+# Modern Gradle Publish Plugin
+
+
+## Goals
+
+- less magic configuration, more explicit APIs
+- provide APIs for everything represented through properties
+- remove or deprecate APIs that are already provided by Gradle
+- continue providing a way to get started with minimal configuration
+
+## New Structure
+
+For compatibility purposes the current `com.vanniktech.maven.publish` plugin
+behavior is kept as it is, with the exception of APIs that we decide to remove. 
+Most of the goals are solved through a set of new plugins for specific 
+platforms. These plugins provide APIs that replace the magic setup mechanisms
+of the current plugin. To avoid having to maintain two implentations the old
+plugin will internally just apply the new plugins and use their public APIs. 
+Over time we can decide if `com.vanniktech.maven.publish` should be deprecated
+or removed.
+
+### Plugins
+
+Specific plugins for each supported project type:
+- `com.vanniktech.maven.publish.java`
+- `com.vanniktech.maven.publish.gradle`
+- `com.vanniktech.maven.publish.android`
+- `com.vanniktech.maven.publish.kotlin.jvm`
+- `com.vanniktech.maven.publish.kotlin.js`
+- `com.vanniktech.maven.publish.kotlin.mpp`
+
+The main motivation for separating these, is to provide specific APIs for each project
+without running into race conditions when something like `plugins.withId` is used or 
+requiring ordering. Another reason is that it allows us to nicely separate the code. 
+
+On top of that there will be a `com.vanniktech.maven.publish.base` plugin that 
+gets applied by all of these and the existing `com.vanniktech.maven.publish` as 
+described above.
+
+#### `com.vanniktech.maven.publish.base`
+
+- applies `maven-publish` Gradle plugin
+- provides the API that is shared for all project types
+- can be used for shared configurations using `plugins.withId(...)` in projects with mixed types
+
+```
+mavenPublish {
+  # applies signing plugin and tells Gradle to sign all publications
+  signAllPublications()
```

Comment:
Can we have more options there?

```
// Uses the gradle properties as explained in https://docs.gradle.org/current/userguide/signing_plugin.html#sec:signatory_credentials
signAllPublications()
// Uses in-memory keys which is more convenient to hold in an env variable in CI
signAllPublicationsWithInMemoryKeys(key, password)

// Allow to easily disable signing if developers do not have the secrets or in order to save some time
signingIsEnabled = ... //System.genv(SINGING_KEY) != null


```

Reply:
Avoid signing is important because snapshots doesnt require it and time is gold.

Reply:
I would keep the current behavior of not trying to sign when the current versions ends with `-SNAPSHOT`. If you don't want to sign non snapshot artifacts then you could also just do `if (conditionForSigning) signAllPublications()`, I don't think we need to offer an extra option here.

As for `signAllPublicationsWithInMemoryKeys`, I'm a bit split on this. It would be a bit shorter
```
mavenPublish {
  signAllPublications()
}

signing {
  if (project.hasProperty('SIGNING_PRIVATE_KEY') && project.hasProperty('SIGNING_PASSWORD')) {
    useInMemoryPgpKeys(project.getProperty('SIGNING_PRIVATE_KEY'), project.getProperty('SIGNING_PASSWORD'))
  }
}
```
vs 
```
mavenPublish {
  if (project.hasProperty('SIGNING_PRIVATE_KEY') && project.hasProperty('SIGNING_PASSWORD')) {
    signAllPublicationsWithInMemoryKeys(project.getProperty('SIGNING_PRIVATE_KEY'), project.getProperty('SIGNING_PASSWORD'))
  }
}
```
but also not too much. I could see value if we just have expected property names and the API would be `signAllPublicationsWithInMemoryKeys()` (again this would only make signing required for non SNAPSHOTS). While we are going away from Gradle properties they are still the best solution for handling credentials which this falls under. I have also opened an issue with Gradle to support the credentials API for signing https://github.com/gradle/gradle/issues/16056

Reply:
> If you don't want to sign non snapshot artifacts then you could also just do `if (conditionForSigning) signAllPublications()`, I don't think we need to offer an extra option here.

Agreed that putting the `signAllPublications` under a `if` condition is simpler, no need for `signingIsEnabled()` 👍 

> While we are going away from Gradle properties they are still the best solution for handling credentials which this falls under.

I'd have to respectfully disagree here. I find environment variables to be easier in many circumstances. Mainly because they are super easy to setup in CI as well as on the dev machine.

Reply:
You can create gradle properties by prefixing an environment variable with `ORG_GRADLE_PROJECT_`. For example you can have `ORG_GRADLE_PROJECT_mavenCentralRepositoryUsername` as an env var for the `mavenCentralRepositoryUsername` property

Reply:
We're doing that here for example https://github.com/vanniktech/gradle-maven-publish-plugin/blob/master/.github/workflows/publish-release.yml#L25-L28

Reply:
TBH, this is exactly the part that gets me confused every time 😅 . Gradle properties come with a lot of cognitive load (project vs system, ~/.gradle.properties or project/gradle.properties or env variables, does it have to be upper or lower case, etc, etc.....). I've read https://docs.gradle.org/current/userguide/build_environment.html multiple times and I still can't wrap my head around all the possibilities opened by gradle.properties. That why I usually prefer to avoid them. 

All in all, I think I'd like this plugin to be opiniated and hide the overwhelming options providing by Gradle under a simple API. I think having an api that takes two string parameters (armoured key + password) would be a nice default. That wouldn't prevent someone who wants to sign on a Yubikey to go read the Gradle Signing documentation and configure signing manually using gpg but I think that'd make it easier for beginners.

Reply:
Setting the sonatype username and password already happens through Gradle properties though (using Gradle's credential API). Would adding 2 more properties for in memory signing really make a difference? The default not in memory signing also uses 3 Gradle properties. I'll open a PR today we can discuss more there